File: zenspider/2017/23.rb
#!/usr/bin/env ruby -w

require "prime"

class String
  def to_v
    Integer(self) rescue self.to_sym
  end
end

class Problem23a
  attr_accessor :register, :program, :pc, :counter

  def initialize
    self.register = Hash.new { |h,k| raise "Unknown register: #{k}" }
    [:a, :b, :c, :d, :e, :f, :g, :h].each do |name|
      register[name] = 0
    end

    self.pc = 0
    self.counter = 0
  end

  def parse input
    self.program = input.lines.map { |line|
      case line
      when /set (\S+) (\S+)/ then # set X Y: sets register X to value Y
        x, y = $1.to_v, $2.to_v
        if y.is_a? Integer then
          ->() { register[x] = y;            step }
        else
          ->() { register[x] = register[y];  step }
        end
      when /sub (\S+) (\S+)/ then # sub X Y: subtracts register X by value Y
        x, y = $1.to_v, $2.to_v

        if y.is_a? Integer then
          ->() { register[x] -= y;           step }
        else
          ->() { register[x] -= register[y]; step }
        end
      when /mul (\S+) (\S+)/ then # mul X Y: multiplies register X by value Y
        x, y = $1.to_v, $2.to_v

        if y.is_a? Integer then
          ->() { register[x] *= y;           count; step }
        else
          ->() { register[x] *= register[y]; count; step }
        end
      when /jnz (\S+) (\S+)/ then # jnz X Y: pc += value Y, if the value X != 0
        x, y = $1.to_v, $2.to_v

        if x.is_a? Integer then
          if x != 0 then
            ->() { step y }
          else
            ->() { step 1 }
          end
        else
          if x.is_a? Integer then
            ->() { step (x.zero?           ? 1 : y) }
          else
            ->() { step (register[x].zero? ? 1 : y) }
          end
        end
      else
        warn "unparsed: #{line.chomp}"
      end
    }
  end

  def step n = 1
    self.pc += n
  end

  def count
    self.counter += 1
  end

  def run input
    parse input

    old_pc = self.pc
    inst = program[pc]
    begin
      inst.call
      old_pc = self.pc
    end while inst = program[old_pc]

    self
  end
end

class Problem23b < Problem23a
  def initialize
    super
    register[:a] = 1
  end

  def run
    b = 57 * 100 + 100_000
    c = b + 17_000
    h = 0

    (b..c).step(17).each do |bb|
      h += 1 unless bb.prime?
    end

    h
  end
end

if __FILE__ == $0 then
  if ARGV.empty? then
    require "minitest/autorun"

    class Test23 < Minitest::Test
      def test_set
        cpu = Problem23a.new

        assert_equal  0, cpu.register[:a]
        cpu.run "set a 42"
        assert_equal 42, cpu.register[:a]
        assert_equal 1, cpu.pc
      end

      def test_sub
        cpu = Problem23a.new

        assert_equal  0, cpu.register[:a]
        cpu.run "sub a -42"
        assert_equal 42, cpu.register[:a]
        assert_equal 1, cpu.pc
      end

      def test_mul
        cpu = Problem23a.new
        cpu.register[:a] = 2

        cpu.run "mul a 21"
        assert_equal 42, cpu.register[:a]
        assert_equal 1, cpu.pc
        assert_equal 1, cpu.counter
      end

      def test_jnz_true
        cpu = Problem23a.new
        cpu.register[:a] = 1

        cpu.run "jnz a 42"
        assert_equal 42, cpu.pc
      end

      def test_jnz_true_const
        cpu = Problem23a.new
        cpu.register[:a] = 1

        cpu.run "jnz 1 42"
        assert_equal 42, cpu.pc
      end

      def test_jnz_false
        cpu = Problem23a.new
        cpu.register[:a] = 0 # redundant

        cpu.run "jnz a 42"
        assert_equal 1, cpu.pc
      end

      def test_b
        skip
      end
    end
  else
    input = ARGF.read.chomp
    p Problem23a.new.run(input).counter
    p Problem23b.new.run
  end
end

File: zenspider/2017/23.hs
#!/usr/bin/env runhaskell

import Test.HUnit
import System.IO (stderr)
import System.Environment (getEnv)
-- import Data.Char

-- cpu = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

-- data LispVal = Atom String
--              | List [ LispVal ]
--              | DottedList [ LispVal ] LispVal
--              | Number Integer
--              | String String
--              | Bool Bool

data Value = Register Char
           | Int Integer

x :: Value
x = Int 42

y :: Value
y = Register 'a'

data Slot = PC Integer
          | Counter Integer
          | A Integer
          | B Integer
          | C Integer
          | D Integer
          | E Integer
          | F Integer
          | G Integer
          | H Integer

-- Num a => (a, [Char])

-- data CPU :: Integer pc, Integer count, Slot reg => (pc, count, reg, reg, reg, reg, reg, reg, reg, reg)

problem23a :: String -> Int
problem23a _ = 42

problem23b :: String -> Int
problem23b _ = 42

tests :: Test
tests =
  test
    [ 24  ~=? problem23a "1122"
    ------------------------------
    , 24  ~=? problem23b "1212" ]

runTests :: Test -> IO (Counts, Int)
runTests ts = do
  term <- getEnv "TERM"
  let smart = "dumb" /= term
  runTestText (putTextToHandle stderr smart) ts

main :: IO ()
main = do
  _ <- runTests tests
  [input] <- lines <$> readFile "input/23.txt"
  print $ problem23a input
  print $ problem23b input


-- fuuuuck you
-- main = print $ length $ filter (not.isPrime.head) $ chunksOf 17 [b..b+17000] where b = 108100


-- data Ins reg imm
--   = Set {reg :: reg, val :: Either reg imm}
--   | Sub {reg :: reg, val :: Either reg imm}
--   | Mul {reg :: reg, val :: Either reg imm}
--   | Jnz {cnd :: Either reg imm, jmp :: Either reg imm}
-- data State a reg imm = State {pc :: Int, regs :: a reg imm}
--
-- parseVal :: (Read imm) => String -> Either Char imm
-- parseVal (readMaybe -> Just imm) = Right imm
-- parseVal [reg] = Left reg
--
-- parse :: (Read imm) => String -> [Ins Char imm]
-- parse = map (parseIns . words) . lines where
--     parseIns ["set", [reg], parseVal -> val] = Set {..}
--     parseIns ["sub", [reg], parseVal -> val] = Sub {..}
--     parseIns ["mul", [reg], parseVal -> val] = Mul {..}
--     parseIns ["jnz", parseVal -> cnd, parseVal -> jmp] = Jnz {..}
--
-- step :: (IArray a imm, Ix reg, Integral imm) =>
--     [Ins reg imm] -> State a reg imm -> Maybe (State a reg imm)
-- step ins State {pc} | pc < 0 || null (drop pc ins) = Nothing
-- step ins state@State {..} = Just $ case ins !! pc of
--     Set {..} -> State (pc + 1) (regs // [(reg, load val)])
--     Sub {..} -> State (pc + 1) (regs // [(reg, regs ! reg - load val)])
--     Mul {..} -> State (pc + 1) (regs // [(reg, regs ! reg * load val)])
--     Jnz {..}
--       | load cnd == 0 -> state {pc = pc + 1}
--       | otherwise -> state {pc = pc + fromIntegral (load jmp)}
--   where
--     load = either (regs !) id
--
-- iterateMaybe :: (a -> Maybe a) -> Maybe a -> [a]
-- iterateMaybe f = maybe [] $ (:) <*> iterateMaybe f . f
--
-- day23a :: String -> Int
-- day23a input = countMuls $ iterateMaybe (step ins) $ Just state0 where
--     ins = parse input
--     muls = findIndices (\case Mul {} -> True; _ -> False) ins
--     countMuls = length . filter (`elem` muls) . map pc
--     state0 = State {pc = 0, regs = listArray @UArray @Int ('a', 'h') $ repeat 0}

-- pt 2:
-- length $ filter (not . isPrime) [107900, 107917 .. 124900]
--
-- or:

-- stepOptimized :: (IArray a imm, Integral imm) =>
--     [Ins Char imm] -> State a Char imm -> Maybe (State a Char imm)
-- stepOptimized ins State {pc} | pc < 0 = Nothing
-- stepOptimized ins state@State {..} = case drop pc ins of
--     ( Set 'd' (Right 2) :
--       Set 'e' (Right 2) :
--       Set 'g' (Left 'd') :
--       Mul 'g' (Left 'e') :
--       Sub 'g' (Left 'b') :
--       Jnz (Left 'g') (Right 2) :
--       Set 'f' (Right 0) :
--       Sub 'e' (Right (-1)) :
--       Set 'g' (Left 'e') :
--       Sub 'g' (Left 'b') :
--       Jnz (Left 'g') (Right (-8)) :
--       Sub 'd' (Right (-1)) :
--       Set 'g' (Left 'd') :
--       Sub 'g' (Left 'b') :
--       Jnz (Left 'g') (Right (-13)) :
--       _) -> Just . State (pc + 15) $ regs //
--             ( [ ('d', regs ! 'b'), ('e', regs ! 'b'), ('g', 0) ] ++
--               [('f', 0) | not . isPrime . fromIntegral $ regs ! 'b'] )
--     _ -> step ins state
--
-- day23b :: String -> Int
-- day23b input =
--     regs (last . iterateMaybe (stepOptimized ins) $ Just state0) ! 'h' where
--     ins = parse input
--     state0 =
--         State {pc = 0, regs = listArray @UArray @Int ('a', 'h') $ 1 : repeat 0}

-- also: https://github.com/NeilNjae/advent-of-code-17/blob/master/src/advent23/advent23.hs
